lib/ansible/plugins/action/aws_s3.py

```diff
@@ -43,11 +43,14 @@ def run(self, tmp=None, task_vars=None):
             new_module_args = self._task.args.copy()
             if source:
                 source = os.path.expanduser(source)
-                try:
-                    source = self._loader.get_real_file(self._find_needle('files', source))
-                    new_module_args['src'] = source
-                except AnsibleError as e:
-                    raise AnsibleActionFail(to_text(e))
+
+                # For backward compatibility check if the file exists on the remote; it should take precedence
+                if not self._remote_file_exists(source):
+                    try:
+                        source = self._loader.get_real_file(self._find_needle('files', source))
+                        new_module_args['src'] = source
```

Comment:
Do we also need to copy the file from controller to target? I've only ever tried this with controller == target.

Reply:
Hey, @willthames sorry for taking a while to get back to this one. I wasn't sure exactly how it all worked and what the expected behavior is so I took some time to test it out and compare with older versions. It appears the controller file has never been copied so it would be a feature. Since I'd like this to be backported that could be a separate discussion. I have tested with a target that is not the controller, and this fixes uploading a file that exists on the target.

I think I'm going to clarify the error if the file doesn't exist on the remote or the controller because it is a little confusing:
```
"msg": "Could not find or access '/tmp/onlyonremote.txt' on the Ansible Controller.\nIf you are using a module and expect the file to exist on the remote, see the remote_src option"
```
If the file exists on the controller but not the remote I get this message, which seems better:
```
"msg": "Local object for PUT does not exist"
```